### ISSUE TITLE
TTF line height calculation issue

### ISSUE THREAD

**Issue by rh101** [adxeproject/adxe#383]
- adxe version: current

This issue is going to be a little tough to explain, so please bear with me.

The FreeType library bundled with Cocos2d-x (v4 and prior) is v2.5.5, which has a specific bit of code that attempts to correct the calculation of the TrueType font ascender and descender metric values (in `ttdriver.c :: tt_size_request()`).

In later versions of FreeType, that bit of code is in `ttdriver.c :: tt_size_request()` was reverted back to the older values, and the new calculated values are now stored in another location. This reason for this seemed to be due to issues with existing applications/code that relied on the old values before the fix was implemented.

In ADXE, the latest version of FreeType is being used, which means the values of ascender and descender are different to those generated in Cocos2d-x v4.

The code that is generating the line height in ADXE and Cocos2d-x using the ascender and descender values:

```
bool FontFreeType::createFontObject(const std::string &fontName, float fontSize)
{
...
...
    _fontRef = face;
    _lineHeight = static_cast<int>((_fontRef->size->metrics.ascender - _fontRef->size->metrics.descender) >> 6);
...
}
```

As an example, we use the `Arimo-Regular.ttf` font.  For font size 29, the line height is calculated as follows:

In Cocos2d-x/FreeType v2.5.5, `_lineHeight` = 32.

In ADXE/FreeType v2.10.4, `_lineHeight ` = 34.

This issue is noticeable when using multi-line text labels, and it results in bigger gaps between lines.

In v2.10.4 of FreeType, the same values as v2.5.5 for the TrueType calculation are now stored in `((TT_Size)face->size)->metrics`.

So, if you want to fix this issue (if it is an issue for you), then to get the adjusted TrueType values for ascender and descender as they were in FreeType 2.5.5, this is the change that needs to be made:
```
    // store the face globally
    _fontRef = face;
    auto* ttSize = (TT_Size)(_fontRef->size);
    _lineHeight = static_cast<int>((ttSize->metrics->ascender - ttSize->metrics->descender) >> 6);
```

With this change, for `Arimo-Regular.ttf` font, font size 29, the `_lineHeight` = 32, which is correct, and fixes the issue with line spacing.

This will also affect this method:
```
int FontFreeType::getFontAscender() const
{
    return (static_cast<int>(_fontRef->size->metrics.ascender >> 6));
}
```
which would become:
```
int FontFreeType::getFontAscender() const
{
    auto* ttSize = (TT_Size)(_fontRef->size);
    return (static_cast<int>(ttSize->size->metrics->ascender >> 6));
}
```

Have you come across this issue?  

If you want the same output as Cocos2d-x v4/FreeType v2.5.5, then that fix above is needed.  For me personally, the changes above produce the correct multi-line text line spacing output in our project.  I've compared it to what the artists are producing in Adobe Flash, and it's exactly the same with the above fix.

The ascender and descender values of TTF can be inaccurate at times, and sometimes a work-around is added in code-bases to get around these issues.  If the code above is used, then it will affect any project using the current version of ADXE, so it's a tough one to decide whether to apply the change or not.  

I've added this change to my own fork for now, since I require it, but it really is up to you whether you want to make the change to the main branch of the ADXE code.

For more info, here is when the issue was initially raised:
https://lists.nongnu.org/archive/html/freetype-devel/2011-07/msg00022.html

The code fix was first added here (this is what FreeType v2.5.5 has):
https://gitlab.freedesktop.org/freetype/freetype/-/commit/b0962ac34e66052ccfee7996e5468f30d4bd5a72

Then modified/removed here (but the fixed values were stored in the TT_Size structure), and what exists in FreeType 2.10.4:
https://gitlab.freedesktop.org/freetype/freetype/-/commit/bcc74f4dafee25ea89f1d3144646cba7e30f9908

**Comment by halx99** [comment]
I think you can send a pr to mian branch of adxe.

**Comment by rh101** [comment]
Sure, I'll create a PR soon. Probably best to get some feedback from others using ADXE to see how this impacts them.

**Comment by rh101** [comment]
Fixed in #384
